presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveTransactionalTable.java

```diff
@@ -184,19 +187,66 @@ private static String hiveTableProperties(TransactionalTableType transactionalTa
 
     private static void compactTableAndWait(CompactionMode compactMode, String tableName, String partitionString, Duration timeout)
     {
-        assertEquals(getTableCompactions(compactMode, tableName).count(), 0);
+        log.info("Running %s compaction on %s", compactMode, tableName);
+
+        Failsafe.with(
+                new RetryPolicy<>()
+                        .withMaxDuration(java.time.Duration.ofMillis(timeout.toMillis()))
+                        .withMaxAttempts(Integer.MAX_VALUE))  // limited by MaxDuration
+                .onFailure(event -> {
+                    throw new IllegalStateException(format("Could not compact table %s in %d retries", tableName, event.getAttemptCount()), event.getFailure());
+                })
+                .onSuccess(event -> log.info("Finished %s compaction on %s in %s (%d tries)", compactMode, tableName, event.getElapsedTime(), event.getAttemptCount()))
+                .run(() -> tryCompactingTable(compactMode, tableName, partitionString, Duration.valueOf("90s")));
+    }
+
+    private static void tryCompactingTable(CompactionMode compactMode, String tableName, String partitionString, Duration timeout)
+            throws TimeoutException
+    {
+        // Hive's unix_timestamp() returns millis
+        long beforeCompactionStart = (Long) onHive().executeQuery("SELECT unix_timestamp()").row(0).get(0) * 1000;
         onHive().executeQuery(format("ALTER TABLE %s %s COMPACT '%s'", tableName, partitionString, compactMode.name())).getRowsCount();
 
-        // Since we disabled table auto compaction and we checked that there are no compaction to the table
-        // we can assume that every compaction from now on is triggered in this test
-        // and all compaction should complete successfully before proceeding.
-        assertEventually(timeout, () -> {
-            Map<String, String> compaction = getOnlyElement(getTableCompactions(compactMode, tableName)
-                    .collect(toImmutableList()));
+        log.info("Started compactions: %s", getTableCompactions(compactMode, tableName).collect(toImmutableList()));
+
+        long loopStart = System.nanoTime();
+        while (true) {
+            try {
+                // Compaction takes couple of second so there is no need to check state more frequent than 1s
+                Thread.sleep(1000);
+            }
+            catch (InterruptedException e) {
+                throw new RuntimeException(e);
+            }
+
+            // Since we disabled auto compaction for uniquely named table and every compaction is triggered in this test
+            // we can expect that single compaction in given mode should complete before proceeding.
+            Optional<Map<String, String>> startedCompaction = getTableCompactions(compactMode, tableName)
+                    .filter(row -> !row.get("state").equals("initiated"))
```

Comment:
I'd rather have this two-step

1. find _ours_ compaction (`getTableCompactions` + filter by time)
   - it must exists and must be exactly one
2. (sep stmt) verify it's "state" ("initiated" or sth else)

Reply:
I know why I've filtered out initiated compactions as the first step. They don't have start time (it's ` --- `)